vllm/lora/punica.py

```diff
@@ -5,17 +5,36 @@
 import torch
 
 from vllm import _custom_ops as ops
+from vllm.logger import init_logger
+from vllm.lora import native_kernels
+from vllm.utils import is_cpu
+
+logger = init_logger(__name__)
+
+if is_cpu():
+    logger.warning(
+        "The CPU backend does not support custom kernels for LoRA. "
+        "Falling back to unoptimized PyTorch-native implementation, "
+        "which may lead to performance drop.")
+elif torch.cuda.get_device_capability() < (8, 0):
```

Comment:
QQ: Is this implementation compatible with CUDA graphs?

Reply:
This implementation works with `enforce_eager=False`, and the CUDA graphs is captured successfully:
```bash
(VllmWorkerProcess pid=6403) WARNING 06-26 03:11:20 punica.py:20] punica LoRA kernels require compute capability >= 8.0, but you are running on device with compute capability < 8.0. Falling back to unoptimized PyTorch-native implementation, which may lead to performance drop.
(VllmWorkerProcess pid=6403) INFO 06-26 03:11:25 selector.py:142] Cannot use FlashAttention-2 backend for Volta and Turing GPUs.
(VllmWorkerProcess pid=6403) INFO 06-26 03:11:25 selector.py:52] Using XFormers backend.
(VllmWorkerProcess pid=6403) INFO 06-26 03:11:26 multiproc_worker_utils.py:215] Worker ready; awaiting tasks
(VllmWorkerProcess pid=6403) INFO 06-26 03:11:27 utils.py:672] Found nccl from library libnccl.so.2
(VllmWorkerProcess pid=6403) INFO 06-26 03:11:27 pynccl.py:63] vLLM is using nccl==2.20.5
(VllmWorkerProcess pid=6403) INFO 06-26 03:11:27 custom_all_reduce_utils.py:208] reading GPU P2P access cache from /root/.config/vllm/gpu_p2p_access_cache_for_0,1.json
(VllmWorkerProcess pid=6403) WARNING 06-26 03:11:27 custom_all_reduce.py:175] Custom allreduce is disabled because your platform lacks GPU P2P capability or P2P test failed. To silence this warning, specify disable_custom_all_reduce=True explicitly.
(VllmWorkerProcess pid=6403) INFO 06-26 03:11:27 selector.py:142] Cannot use FlashAttention-2 backend for Volta and Turing GPUs.
(VllmWorkerProcess pid=6403) INFO 06-26 03:11:27 selector.py:52] Using XFormers backend.
(VllmWorkerProcess pid=6403) INFO 06-26 03:11:28 weight_utils.py:218] Using model weights format ['*.safetensors']
(VllmWorkerProcess pid=6403) INFO 06-26 03:11:29 model_runner.py:160] Loading model weights took 2.6537 GB
INFO 06-26 03:11:33 distributed_gpu_executor.py:56] # GPU blocks: 1912, # CPU blocks: 1638
(VllmWorkerProcess pid=6403) INFO 06-26 03:11:36 model_runner.py:889] Capturing the model for CUDA graphs. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
(VllmWorkerProcess pid=6403) INFO 06-26 03:11:36 model_runner.py:893] CUDA graphs can take additional 1~3 GiB memory per GPU. If you are running out of memory, consider decreasing `gpu_memory_utilization` or enforcing eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
(VllmWorkerProcess pid=6403) INFO 06-26 03:12:15 model_runner.py:965] Graph capturing finished in 38 secs.
```
So I think this should be compatible with CUDA graphs.